src/function/scalar/date/date_trunc.cpp

```diff
@@ -11,216 +11,304 @@
 
 namespace duckdb {
 
-struct MillenniumTruncOperator {
-	template <class TA, class TR>
-	static inline TR Operation(TA input) {
-		date_t date = Timestamp::GetDate(input);
-		return Timestamp::FromDatetime(Date::FromDate((Date::ExtractYear(date) / 1000) * 1000, 1, 1), dtime_t(0));
-	}
+struct DateTrunc {
+
+	struct MillenniumOperator {
+		template <class TA, class TR>
+		static inline TR Operation(TA input) {
+			date_t date = Timestamp::GetDate(input);
+			return Timestamp::FromDatetime(Date::FromDate((Date::ExtractYear(date) / 1000) * 1000, 1, 1), dtime_t(0));
+		}
+	};
+
+	struct CenturyOperator {
+		template <class TA, class TR>
+		static inline TR Operation(TA input) {
+			date_t date = Timestamp::GetDate(input);
+			return Timestamp::FromDatetime(Date::FromDate((Date::ExtractYear(date) / 100) * 100, 1, 1), dtime_t(0));
+		}
+	};
+
+	struct DecadeOperator {
+		template <class TA, class TR>
+		static inline TR Operation(TA input) {
+			date_t date = Timestamp::GetDate(input);
+			return Timestamp::FromDatetime(Date::FromDate((Date::ExtractYear(date) / 10) * 10, 1, 1), dtime_t(0));
+		}
+	};
+
+	struct YearOperator {
+		template <class TA, class TR>
+		static inline TR Operation(TA input) {
+			date_t date = Timestamp::GetDate(input);
+			return Timestamp::FromDatetime(Date::FromDate(Date::ExtractYear(date), 1, 1), dtime_t(0));
+		}
+	};
+
+	struct QuarterOperator {
+		template <class TA, class TR>
+		static inline TR Operation(TA input) {
+			date_t date = Timestamp::GetDate(input);
+
+			int32_t month = Date::ExtractMonth(date);
+			month = 1 + (((month - 1) / 3) * 3);
+			return Timestamp::FromDatetime(Date::FromDate(Date::ExtractYear(date), month, 1), dtime_t(0));
+		}
+	};
+
+	struct MonthOperator {
+		template <class TA, class TR>
+		static inline TR Operation(TA input) {
+			date_t date = Timestamp::GetDate(input);
+			return Timestamp::FromDatetime(Date::FromDate(Date::ExtractYear(date), Date::ExtractMonth(date), 1),
```

Comment:
Using `Date::Convert` here instead of ExtractYear + ExtractMonth should save some cycles :)

Reply:
Good call - I wasn't looking at that stuff.